In [7]:
import torch
import numpy as np

In [3]:
torch.__version__

'2.1.0'

In [4]:
device = torch.device('cpu')
device

device(type='cpu')

# Let's start

<img src='../ANN/japan.png'/>

In [6]:
'''
m = 5, n = 3(Temp, Rainfall, Humidity)
y = yield of Apples, Orange
'''

'\nm = 5, n = 3(Temp, Rainfall, Humidity)\ny = Apples, Orange\n'

## 1: Specify X and y

In [8]:
# Input (temp, rainfall, humidity)
X_train = np.array([[73, 67, 43], [91, 88, 64], [87, 134, 58], 
                   [102, 43, 37], [69, 96, 70], [73, 67, 43], 
                   [91, 88, 64], [87, 134, 58], [102, 43, 37], 
                   [69, 96, 70], [73, 67, 43], [91, 88, 64], 
                   [87, 134, 58], [102, 43, 37], [69, 96, 70]], 
                  dtype='float32')

# Targets (apples, oranges)
Y_train = np.array([[56, 70], [81, 101], [119, 133], 
                    [22, 37], [103, 119], [56, 70], 
                    [81, 101], [119, 133], [22, 37], 
                    [103, 119], [56, 70], [81, 101], 
                    [119, 133], [22, 37], [103, 119]], 
                   dtype='float32')

In [9]:
# have to convert numpy to torch tensors
# torch tensors are simply numpy version of pytorch
inputs  = torch.from_numpy(X_train)
targets = torch.from_numpy(Y_train)

print(type(inputs))
print(type(targets))

<class 'torch.Tensor'>
<class 'torch.Tensor'>


In [14]:
inputs.shape, targets.shape
inputs.size(), targets.size()

(torch.Size([15, 3]), torch.Size([15, 2]))

## 2. DataLoaders

Remember in ML - we have to do the batch learning ourselve, like mini-batch, stochastic batch. 
Pytorch has a class called DataLoaders that automatically do this for you.

It's optional whether you wanna use it or just use it - don't reinvent the wheel.

In [22]:
# Define a Dataset so that dataloaders understand

from torch.utils.data import TensorDataset

train_dataset = TensorDataset(inputs, targets)

train_dataset[0]
# get the 3 samples from the dataset
# train_dataset[0:3]

(tensor([73., 67., 43.]), tensor([56., 70.]))

In [27]:
from torch.utils.data import DataLoader

batch_size = 3 # for no reason - here I am using a mini-batch of size 3
train_dl = DataLoader(train_dataset, batch_size, shuffle = True, num_workers = 4)

In [ ]:
for x,y in train_dl:
    print("----new batch----")
    print(x) # this will give you "batch size " of x, e.g. 3 sets of x
    print(y) # this will give you "batch size " of y, e.g. 3 sets of y

## 3. Define some layers

Don't be confuse about "Layers". it's simply matrix multiplication. It's true until now.

In [29]:
import torch.nn as nn # nn = neural nework

In [30]:
some_random_layer = nn.Linear(3,2) # 3 = incoming features, 2 = output features

In [33]:
print(some_random_layer.weight) # matrix - randomly initialize --> using standardization methods.
print(some_random_layer.weight.shape)
print(some_random_layer.bias)
print(some_random_layer.bias.shape)

Parameter containing:
tensor([[ 0.5407,  0.1850, -0.3178],
        [ 0.1833,  0.2263,  0.1200]], requires_grad=True)
torch.Size([2, 3])
Parameter containing:
tensor([-0.2637,  0.3207], requires_grad=True)
torch.Size([2])


Note: output features is 2. So, bias has 2.

In [34]:
outputs = some_random_layer(inputs)
# input = (15,3) @ (3,2) = (15, 2)
outputs.shape

torch.Size([15, 2])

In [35]:
# how do chatgpt count their number of parameters - 175b parameters

total_num_of_params = 0
for param in some_random_layer.parameters():
    print(param)
    # total_num_of_params+=

Parameter containing:
tensor([[ 0.5407,  0.1850, -0.3178],
        [ 0.1833,  0.2263,  0.1200]], requires_grad=True)
Parameter containing:
tensor([-0.2637,  0.3207], requires_grad=True)


print(param)

    Parameter containing: [the first parameters, w = 5]
    tensor([[ 0.5407,  0.1850, -0.3178],
            [ 0.1833,  0.2263,  0.1200]], requires_grad=True) 
    
    Parameter containing: [the second parameters, bias = 2]
    tensor([-0.2637,  0.3207], requires_grad=True)

In [37]:
for param in some_random_layer.parameters():
    print(param.numel())

6
2


In [39]:
total_num_of_params = 0
for param in some_random_layer.parameters():
    total_num_of_params += param.numel()
    
print(total_num_of_params)

8


## 4. Define Loss function